static/app/components/tours/tourContext.tsx

```diff
@@ -0,0 +1,188 @@
+import {type Dispatch, type Reducer, useCallback, useMemo, useRef} from 'react';
+import {useReducer} from 'react';
+
+export type TourEnumType = string | number;
+
+export interface TourStep<T extends TourEnumType> {
+  /**
+   * The element to focus on when the tour step is active.
+   * This is usually set within the TourElement component, which wraps the focused element.
+   */
+  element: HTMLElement | null;
+  /**
+   * Unique ID for the tour step.
+   */
+  id: T;
+}
+
+type TourStartAction<T extends TourEnumType> = {
+  type: 'START_TOUR';
+  stepId?: T;
+};
+type TourNextStepAction = {
+  type: 'NEXT_STEP';
+};
+type TourPreviousStepAction = {
+  type: 'PREVIOUS_STEP';
+};
+type TourEndAction = {
+  type: 'END_TOUR';
+};
+type TourSetRegistrationAction = {
+  isRegistered: boolean;
+  type: 'SET_REGISTRATION';
+};
+
+export type TourAction<T extends TourEnumType> =
+  | TourStartAction<T>
+  | TourNextStepAction
+  | TourPreviousStepAction
+  | TourEndAction
+  | TourSetRegistrationAction;
+
+export interface TourState<T extends TourEnumType> {
+  /**
+   * The current active tour step. If this is null, the tour is not active.
+   */
+  currentStepId: TourStep<T>['id'] | null;
+  /**
+   * Whether the tour is available to the user. Should be set by flags or other conditions.
+   */
+  isAvailable: boolean;
+  /**
+   * Whether every step of the tour has been registered in the DOM.
+   */
+  isRegistered: boolean;
+  /**
+   * The ordered step IDs. Declared once when the provider is initialized.
+   */
+  orderedStepIds: readonly T[];
+}
+
+type TourRegistry<T extends TourEnumType> = {
+  [key in T]: TourStep<T> | null;
+};
+
+export function useTourReducer<T extends TourEnumType>(
+  initialState: TourState<T>
+): TourContextType<T> {
+  const {orderedStepIds} = initialState;
+
+  const reducer: Reducer<TourState<T>, TourAction<T>> = useCallback(
+    (state, action) => {
+      const completeTourState = {...state, currentStepId: null};
+      switch (action.type) {
```

Comment:
A good way to ensure a reducer stays pure is to define it outside of react, that also has a nice benefit of making a reducer easily testable as you can directly call it without going through the react layer.